cluster-autoscaler/cloudprovider/gce/gce_manager.go

```diff
@@ -376,6 +376,68 @@ func (m *GceManager) buildNodeFromTemplate(mig *Mig, template *gce.InstanceTempl
 	return &node, nil
 }
 
+func (m *GceManager) buildNodeFromAutoprovisioningSpec(mig *Mig) (*apiv1.Node, error) {
+
+	if mig.spec == nil {
+		return nil, fmt.Errorf("no spec in mig %s", mig.Name)
+	}
+
+	node := apiv1.Node{}
+	nodeName := fmt.Sprintf("%s-autoprovisioned-template-%d", mig.Name, rand.Int63())
+
+	node.ObjectMeta = metav1.ObjectMeta{
+		Name:     nodeName,
+		SelfLink: fmt.Sprintf("/api/v1/nodes/%s", nodeName),
+		Labels:   map[string]string{},
+	}
+	node.Status = apiv1.NodeStatus{
+		Capacity: apiv1.ResourceList{},
+	}
+	// TODO: get a real value.
+	// TODO: handle GPU
+	node.Status.Capacity[apiv1.ResourcePods] = *resource.NewQuantity(110, resource.DecimalSI)
+
+	cpu, mem, err := m.getCpuAndMemoryForMachineType(mig.spec.machineType)
+	if err != nil {
+		return nil, err
+	}
+	node.Status.Capacity[apiv1.ResourceCPU] = *resource.NewQuantity(cpu, resource.DecimalSI)
+	node.Status.Capacity[apiv1.ResourceMemory] = *resource.NewQuantity(mem, resource.DecimalSI)
+
+	// TODO: use proper allocatable!!
+	node.Status.Allocatable = node.Status.Capacity
+
+	labels, err := buildLablesForAutoprovisionedMig(mig, nodeName)
+	if err != nil {
+		return nil, err
+	}
+	node.Labels = labels
+	// Ready status
+	node.Status.Conditions = cloudprovider.BuildReadyConditions()
+	return &node, nil
```

Comment:
There is a lot of overlap between this method and buildNodeFromTemplate and a lot of identical TODOs, hinting at even more common logic in future. I'm somewhat worried at some point we'll fix something in one place and forget to fix it in the other. Maybe it would be worth to try and extract more common logic?

Reply:
Done.